### PR TITLE
Fix incorrect cookie split

### LINKED ISSUE #930: ArmeriaHttpUtil.toArmeria breakes cookie into linked list entry which will lead only the first cookie entry can be obtained
Netty's `DefaultHeaders` will only get the value of first entry after head node:
https://github.com/netty/netty/blob/4.1/codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java#L117

But in armeria's header conversion, value of cookie header is broken into linked list entries.
https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/internal/ArmeriaHttpUtil.java#L329

This will lead armera's cookie can only obtain the first cookie entry's value.

Example:
```
cookie: a=1; b=2;

But only a=1 will be obtained from HttpRequest#headers().get("cookie").
```

### PR DESCRIPTION
**Pull request by trustin**
Motivation:

The pull request #929 introduced the violation of RFC 7540 section
8.1.2.5:

> If there are multiple Cookie header fields after decompression, these
> MUST be concatenated into a single octet string ..

Modifications:

- Merge and split cookies in a correct and robust way using StringJoiner
  and Splitter
- Inline Http2ToHttpHeaderTranslator which is used only in a single place

Result:

- Fixes #930

### CODE REVIEW

**Comment by trustin** [comment]
@southernkasaist Please review as well.

File: core/src/main/java/com/linecorp/armeria/internal/ArmeriaHttpUtil.java [hunk 1]
```diff
@@ -319,34 +339,22 @@ public static void toArmeria(io.netty.handler.codec.http.HttpHeaders inHeaders,
...
-
-                        if (start < value.length()) {
-                            out.add(HttpHeaderNames.COOKIE, value.substring(start));
-                        }
-                    } else {
-                        out.add(HttpHeaderNames.COOKIE, value);
+                    if (cookieJoiner == null) {
+                        cookieJoiner = new StringJoiner(COOKIE_SEPARATOR);
                     }
+                    COOKIE_SPLITTER.split(value).forEach(cookieJoiner::add);
```
Comment (anuraaga) [inline core/src/main/java/com/linecorp/armeria/internal/ArmeriaHttpUtil.java]: Shouldn't this conversion be in `toAmeria(Http2Headers` and not here? AFAIU, the spec is only related to HTTP/2 while a similar conversion is not needed for HTTP/1.

File: core/src/test/java/com/linecorp/armeria/internal/ArmeriaHttpUtilTest.java [hunk 2]
```diff
@@ -48,7 +53,38 @@ public void testConcatPaths() throws Exception {
...
+        in.add(HttpHeaderNames.COOKIE, "i=j");
+        in.add(HttpHeaderNames.COOKIE, "k=l;");
+
+        final Http2Headers out = toNettyHttp2(in);
+        assertThat(out.getAll(HttpHeaderNames.COOKIE))
+                .containsExactly("a=b", "c=d", "e=f", "g=h", "i=j", "k=l");
+    }
+
+    @Test
+    public void inboundCookiesMustBeMerged() {
```
Comment (anuraaga) [inline core/src/test/java/com/linecorp/armeria/internal/ArmeriaHttpUtilTest.java]: Neet test for http2 input as well.

Also I think you take special care to handle Cookie, Cookie, Content-Type, Cookie etc (mixed up headers) would be good to add that to the tests

Comment (trustin) [reply core/src/test/java/com/linecorp/armeria/internal/ArmeriaHttpUtilTest.java]: Indeed. Fixed as suggested.

File: core/src/main/java/com/linecorp/armeria/internal/ArmeriaHttpUtil.java (hunk 1 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/internal/ArmeriaHttpUtil.java]: Right. I added it just to make it more robust so it handles a poorly written peer. Should we just accept only the first one?

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/internal/ArmeriaHttpUtil.java]: .. and, yes, the conversion should also be in `toArmeria(Http2Headers)`. Added there as well.

Comment (anuraaga) [reply core/src/main/java/com/linecorp/armeria/internal/ArmeriaHttpUtil.java]: As a user I have some expectation that mutiple cookie headers would not be handled specially for HTTP/1 but think either way works. I'd just follow whatever tomcat etc does not sure what that is though :)

**Comment by southernkasaist** [comment]
(awthanks)

**Pull request merged**